### DISCUSSION TITLE
How to do retries after timeouts?

### DISCUSSION THREAD

**Discussion by sykim4** [sindresorhus/ky#279] [Q&A] [answered]
It could be that my ky fetch isn't configured correctly but I haven't seen retry attempts after timeouts. Is this something that is supported with ky/ky-universal?

**Comment by sykim4** [comment]
actually it looks like the calculateRetryDelay, we are excluding timeouts:
`if (this._retryCount < this._options.retry.limit && !(error instanceof TimeoutError)) `

 is there a reason in particular why timeouts are excluded from retries?

**Comment by sindresorhus** [comment]
Why? You can just set a longer timeout instead.

**Reply by steveoh** [reply]
I think hedging is a good reason.

> A hedging strategy enables the re-execution of a request if the previous execution takes too long. Implementing a hedging strategy can boost the overall responsiveness of the system. However, it's essential to note that this improvement comes at the cost of increased resource utilization.

**Comment by sholladay** [comment, marked as answer]
Timeout means to give up on the request, e.g. to avoid having users stare at a loading spinner forever. I can see how this could be a little confusing since if the request seems to be hanging, then one could argue it is in a failure state and we retry other failure states. However, the difference is that in the other cases we have a definitive result, whereas a timeout is more like a judgement or prediction without any contextual information. Retrying a request just because it's taking a long time is probably not advisable in most cases. The server might be overloaded and retries would exacerbate the problem. The internet connection might be slow and retries would probably not improve the overall throughput (given that the request would be restarted from scratch).

What I would recommend is setting a relatively long timeout like 30 seconds (or even disable timeouts altogether) and rely on the browser to report failures which Ky will retry. If you're still experiencing hanging requests, it would be good to investigate a little further to determine what is causing it.

**Reply by steveoh** [reply]
I think that makes sense when ky was browser only. But now that it also works in node, I think a fresh look needs to be taken. On the server side we can introduce circuit breakers and other techniques to handle some of your concerns.

**Comment by sykim4** [comment]
that makes sense to me. appreciate the explanation. i'll close this inquiry!

**Comment by jvitormelo** [comment]
Hi, I need to retry after a timeout, is there any built-in method?

**Reply by sholladay** [reply]
You can try/catch the request, check if a caught error is an instance of `TimeoutError`, and if it is, call Ky again with the same arguments.

There's currently no built-in method to do that. PR welcome if someone wants to implement one. It should either require the user to call it manually or it should only trigger for "retriable" requests.

But as I said above, retrying timed out requests can be problematic. Just something to keep in mind.

**Comment by clicktodev** [comment]
resolved by #767
